Corrections to :TSUpdate / :TSInstall and support bang operator

smhc opened the pull request:
List of enhancements:

- :TSUpdate does not blindly attempt re-install of a parser if it is up-to-date, behaves as expected
- :TSUpdate supports multiple params, e.g :TSUpdate cpp bash java
- :TSUpdateSync command to update synchronously
- :TSInstall and :TSInstallSync no longer prompt if a parser is already installed
- :TSInstall and :TSInstallSync support the bang / ! operator to force re-installation
- Doco updates

The main driver for this is to prevent prompting and allow the headless maintenance of parsers. i.e to install and keep them up to date via installation scripts (e.g Plugin post installation steps, docker container creation etc that need synchronous, headless -and deterministic behaviour). The current set of functionality is hindered by prompting or reinstalling unnecessarily.

Hopefully this can be accepted, or at least similar functionality incorporated.

resolves: https://github.com/nvim-treesitter/nvim-treesitter/issues/1408

theHamsta commented on lua/nvim-treesitter/install.lua:
```diff
@@ -302,100 +298,119 @@ local function install_lang(lang, ask_reinstall, cache_folder, install_folder, w
-      shell.select_rm_file_cmd(parser_lib, "Uninstalling parser for "..lang)
-    }
-    M.iter_cmd(command_list, 1, lang, 'Treesitter parser for '..lang..' has been uninstalled')
+      local parser_lib = install_dir..path_sep..lang..".so"
```
Can you use `join_path` from the `utils.lua` module here?

smhc replied:
possibly? This was taken from moving around existing code so I'm hesitant to rework it under this PR.

theHamsta replied:
@smhc can you remember to do this in a follow up PR?

smhc replied:
Honestly I am unlikely to get to it - it is code I didn't touch, it just got indented. With my follow up changes it should actually show the line as unchanged. I'm not a lua expert to be in a position to clean up based on best practices.

stsewd commented on lua/nvim-treesitter/install.lua:
```diff
@@ -302,100 +298,119 @@ local function install_lang(lang, ask_reinstall, cache_folder, install_folder, w
+        skipped = skipped + install_lang(bang, lang, cache_folder, install_folder, with_sync, generate_from_grammar)
+      end
+      if skipped > 0 then
+        print('Parsers already installed. Use :TSInstall! to force reinstallation.')
```
I think we should print this message on each parser that is already installed, or print the whole list here, otherwise isn't clear which parsers are already installed. Also, kind of having the input asking for re-installation was nice, maybe we can leave it since now is possible to avoid the prompt with the bang? We could make the `ask_reinstall` a global option.

smhc replied:
Listing them all becomes quite verbose when they are all installed and up to date. Imagine the scenario you have a post update hook in :PlugUpdate which calls :TSUpdate - it can be quite a lot of noise. We could concatenate a single-line list, but imho people are mainly interested in the ones that got updated, not the ones that were ignored.

And certainly don't want to prompt whether to overwrite for every parser in that scenario. Simply running :PlugUpdate causes the user to have to answer ~30 or so y/n prompts that they have no real way of giving a proper answer and just wanted to get things updated.

I don't think the installation/updates should be interactive in any scenario. It's better for scripting/deployment/installation, more consistent with other commands and is a better user experience.

imho there's just a few scenarios

- initial install of parsers, non-interactive
- update/sync of existing parsers, non-interactive
- clean re-install of parsers, non-interactive

I can't imagine anyone actually setting the 'ask_reinstall' option - seems like a nvim-treesitter developer option only. Although potentially the :TSInstallSync or :TSUpdateSync could be non-interactive while the :TSInstall / :TSUpdate could be interactive.

Note the :TSInstall and :TSIntsallSync commands are fairly redundant now given :TSUpdate will install if its missing. The only benefit for :TSInstall! over :TSUpdate is to allow a forced re-install via the bang attribute.

theHamsta replied:
Can you change this to `skipped..'parsers already installed'`?

theHamsta replied:
I'm also rather in favor of the current behavior asking reinstall. Non-interactive workflows can always use the lua functions directly.

theHamsta commented on lua/nvim-treesitter/install.lua:
```diff
@@ -50,6 +50,10 @@ local function get_installed_revision(lang)
 
+local function is_installed(lang)
+  return #api.nvim_get_runtime_file('parser/'..lang..'.so', false) > 0
+end
```
Isn't this the same as `has_parser` or do you explicitly want to circumvent any caching with own local implementation?

theHamsta commented on lua/nvim-treesitter/utils.lua:
```diff
@@ -6,7 +6,7 @@ local M = {}
 function M.setup_commands(mod, commands)
   for command_name, def in pairs(commands) do
-    local call_fn = string.format("lua require'nvim-treesitter.%s'.commands.%s.run(<f-args>)", mod, command_name)
+    local call_fn = string.format("lua require'nvim-treesitter.%s'.commands.%s.run('<bang>')(<f-args>)", mod, command_name)
```
Wow, that's a bit of a breaking change when we would do that as also other nvim-treesitter modules depend on that. Wouldn't `.commands['%s<bang>']` work also? Then, modules could just opt-in to by providing a `['install!']` function.

theHamsta commented:
This is a really nice feature! Sorry for the long delay in review. I hope you're still working on that PR!

theHamsta commented:
Another idea for a follow up PR is to allow `:TSUninstall python cpp` to work

theHamsta reviewed (changes requested):
My only doubt is the breaking change in `setup_commands`. I don't think it is quite intuitive that `run` is always returning a function. I'd prefer when a command table would provide a `run` and optionally a `['run!']` field for the banged version.

theHamsta commented:
+1 for the branch name :smile:

smhc commented:
> Another idea for a follow up PR is to allow `:TSUninstall python cpp` to work

I've added this in the latest changes.

> My only doubt is the breaking change in setup_commands.

I've reworked this using your suggestion to have a 'run!' command which should be backwards compatible.

smhc commented:
Also the :TSUpdate command remains prompt-less, even when passing in a specific parser. This differs from :TSInstall which will prompt. 

I prefer this behaviour as I want a promptless way of doing: "install this list of parsers if they don't exist, otherwise update them". I can't use :TSInstall as it will result in a prompt, and I couldn't use the old version of :TSUpdate as it didn't support installing or a list of parsers.

ergo I believe using ":TSUpdateSync cpp bash java" is the preferred post install/update hook for vim-plug as it will do as the user intended - whether its a fresh install, upgrade, non-interactive install etc.

theHamsta reviewed (approved):
Looks good!

One thought, the behavior that you want to have sounds like the one achieved by the `ensure_install` option. Maybe add a `:EnsureInstalled cpp python` command (although this would be pretty much `:TSUpdate` now that it accepts options

theHamsta commented:
Can you fix the lint failures? Then, we can merge.

smhc commented:
Ok, I've hopefully fixed the linting issues.

> One thought, the behavior that you want to have sounds like the one achieved by the ensure_install option. Maybe add a :EnsureInstalled cpp python command (although this would be pretty much :TSUpdate now that it accepts options

Yes agreed.  The 'ensure install' stuff is similar to my requirements. But afaik it didn't let me do it synchronously during 'installation' - it's run on first start instead. From what I can see it also installs all/maintained and takes an 'ignore' list - which might work ok but not quite how I wanted to use it. I suppose I could have written a different hook that dug into the bowels of the lua modules to call things directly.

My use case is regular deployments to clean dockerised environments where the plugin installation is done post container creation rather than being part of the docker image. I am still trying to determine the best way to deploy pre-built .so files from a network store as I am not keen on the compilation on every deployment - that may be something I simply roll myself.

theHamsta commented:
@smhc I was just about naming. A `:TSEnsureInstalledSync` would describe your desired behavior best. It could be an alias for `:TSUpdateSync`. But `:TSUpdateSync` should do it for now too.
